core/config/config_helper.hpp

```diff
@@ -0,0 +1,171 @@
+// SPDX-FileCopyrightText: 2017 - 2024 The Ginkgo authors
+//
+// SPDX-License-Identifier: BSD-3-Clause
+
+#ifndef GKO_CORE_CONFIG_CONFIG_HELPER_HPP_
+#define GKO_CORE_CONFIG_CONFIG_HELPER_HPP_
+
+
+#include <string>
+#include <type_traits>
+
+
+#include <ginkgo/core/base/exception_helpers.hpp>
+#include <ginkgo/core/base/lin_op.hpp>
+#include <ginkgo/core/base/math.hpp>
+#include <ginkgo/core/config/config.hpp>
+#include <ginkgo/core/config/registry.hpp>
+#include <ginkgo/core/solver/solver_base.hpp>
+#include <ginkgo/core/stop/criterion.hpp>
+
+
+#include "core/config/registry_accessor.hpp"
+
+
+namespace gko {
+namespace config {
+
+
+/**
+ * LinOpFactoryType enum is to avoid forward declaration, linopfactory header,
+ * two template versions of parse
+ */
+enum class LinOpFactoryType : int { Cg = 0 };
+
+
+/**
+ * It is only an intermediate step after dispatching the class base type. Each
+ * implementation needs to deal with the template selection.
+ */
+template <LinOpFactoryType flag>
+deferred_factory_parameter<gko::LinOpFactory> parse(
+    const pnode& config, const registry& context,
+    const type_descriptor& td = make_type_descriptor<>());
```

Comment:
It will be more clear in another pr.
for example, in solver config pr, it will be instantiated in core/config/solver_config.cpp. I collect them in different file if they are in different namespace.